etc/extract-doc-examples.py

```diff
@@ -0,0 +1,366 @@
+
+    Args:
+        folder_path (string): Path to the folder to process
+        recursive (bool): Whether to process subdirectories recursively
+    """
+    folder = Path(folder_path)
+
+    if not folder.exists():
+        print(f"Error: Directory '{folder_path}' does not exist.")
+        sys.exit(1)
+
+    if not folder.is_dir():
+        print(f"Error: '{folder_path}' is not a directory.")
+        sys.exit(1)
+
+    print(f"Searching for code blocks in folder: {folder.absolute()}")
+    if recursive:
+        print("Processing subdirectories recursively...")
+    print("=" * 60)
+
+    # Get files
+    if recursive:
+        files = folder.rglob("*")
+    else:
+        files = folder.glob("*")
+
+    # Filter
+    files = [f for f in files if f.is_file()]
+
+    if len(files) == 0:
+        print("No files found in the specified directory.")
+        return
+
+    total_blocks = 0
+
+    try:
+        with open(TEST_FILEPATH, "w") as testfile:
+            testfile.write(f"{HEADER_TEXT}\n")
+            testfile.write("namespace libsemigroups {\n")  # Open namespace
+
+            for file_path in sorted(files):
+                # Extract example code
+                if (Path(file_path).name in exclude) or (file_path in exclude):
+                    code_blocks = []
+                    __bold(
+                        f"Note: Excluded {
+                            file_path
+                        } from example code block extraction."
+                    )
+                else:
+                    code_blocks = extract_code_blocks(file_path)
+
+                # Write to cpp test file
+                if code_blocks:
+                    for i, block in enumerate(code_blocks, 1):
+                        testfile.write(
+                            f'// {file_path.relative_to(folder)}: Line {
+                                block["start_line"]
+                            }\nLIBSEMIGROUPS_TEST_CASE("docs", "{
+                                "{:03}".format(total_blocks)
+                            }", "{
+                                file_path.relative_to(folder)
+                            }", "[docs][quick]") {{\n'
+                        )
+                        if block["content"].strip():
+                            testfile.write("    " + block["content"])
+                        else:
+                            testfile.write("// ~ empty code block ~")
+                        testfile.write("\n}\n\n")
+                        total_blocks += 1
+
+            testfile.write("\n}")  # Close namespace
+    except IOError as e:
+        print(f"Could not write to test file: {e}")
+
+    if total_blocks == 0:
+        print("No code blocks found in any files.")
+    else:
+        print(f"Total code blocks found: {total_blocks}")
+
+
+########################################################################
+# File Formatting
+#########################################################################
```

Comment:
Rather than having this formatting section, it might be better to add a recipe at the bottom of `Makefile.in` that generates the doc tests, and then formats the generated files. Something like
```makefile
doc-tests:
	@echo "Extracting code blocks to test. . .";
	python3 etc/extract-doc-examples.py
    clang-format-15 tests/test-docs-code-exampes.cpp
```

This way, you can still run this python file without a formatter, and then format it yourself. Helpful if you have your `clang-format` installed under a different name/alias.

Reply:
Ah my bad. It should be added to `Makefile.am`. `Makefile.in` gets generated from `Makefile.am` when `./configure` is run.